scRNA-seq-advanced/02-celltype_assignment.Rmd

```diff
@@ -0,0 +1,465 @@
+
+So let's separate the training from the fitting.
+First fit the model.
+
+```{r}
+singler_finemodel <- SingleR::trainSingleR(
+  blueprint_encode,
+  labels = blueprint_encode$label.fine,
+  # use DE to select genes (default)
+  genes = "de", 
+  # only use genes in the sce object
+  restrict = rownames(sce),
+  BPPARAM = BiocParallel::MulticoreParam(4)
+)
+```
+
+Now classify the cells based on that model.
+
+```{r}
+singler_result_fine <- SingleR::classifySingleR(
+  sce,
+  singler_finemodel,
+  BPPARAM = BiocParallel::MulticoreParam(4)
+)
+```
+
+
+What labels were assigned, and how many?
+
+```{r}
+table(singler_result_fine$labels)
+```
+
+```{r}
+sce$celltype_fine <- singler_result_fine$pruned.labels
+scater::plotUMAP(sce, colour_by = "celltype_fine")
+```
+
+That's a pretty messy plot.
+Let's collapse a bit more manually this time. 
+
+```{r}
+collapsed_labels <- singler_result_fine$pruned.labels %>%
+  forcats::fct_collapse(
+    "HSC" = c("CLP","CMP", "GMP", "HSC", "Megakaryocytes", "MEP", "MPP"),
+    "B-cells" = c("Class-switched memory B-cells", "Memory B-cells", "naive B-cells", "Plasma cells")
+  ) %>%
+  # Sort all CD4+ labels together at the start
+  forcats::fct_relevel("CD4+ T-cells", "CD4+ Tcm", "CD4+ Tem", "Tregs")
+```
+
+Now we plot.
+
+```{r}
+sce$celltype_collapsed <- collapsed_labels
+scater::plotUMAP(sce, colour_by = "celltype_collapsed")
+```
+
+
+### Heatmap of cell types & clusters
+
+
+Let's look at how the cell type assignments we have compare to the clusters we found from the data alone.
+
+```{r}
+type_cluster_tab <- table(sce$celltype_fine, sce$nn_cluster)
+
+# Normalize by cluster size (columns)
+type_cluster_tab <- apply(
+  type_cluster_tab, 
+  2, # apply function to columns
+  function(x){x/sum(x)} # function to apply
+)
+
+type_cluster_tab <- scale(type_cluster_tab, 
+                          center = FALSE, 
+                          scale = colSums(type_cluster_tab))
+
+# plot with pheatmap
+pheatmap::pheatmap(type_cluster_tab)
+```
+
+
+
+### Classifying by clusters
+
+This section inspired by the fact that `SingleR()` has a `clusters` argument.
+
+First we will make a new matrix where we sum the counts across cells that are from the same type according to our clustering.
+The idea here is that we can reduce noise by combining a bunch of cells.
+
+Note that in this case we are using raw counts. 
+`SingleR` is non-parametric, so it shouldn't matter! 
+(And adding log values felt bad.)
+
+```{r}
+sce_cluster_mat <- DelayedArray::colsum(counts(sce), sce$nn_cluster)
+dim(sce_cluster_mat)
+```
+
+```{r}
+singler_result_cluster <- SingleR::classifySingleR(
+  sce_cluster_mat, 
+  singler_finemodel
+  )
+```
+
+```{r}
+singler_result_cluster$labels
+```
+Add cluster labels to individual cells.
+
+```{r}
+sce$celltype_cluster <- singler_result_cluster$labels[sce$nn_cluster]
+```
+
+Plot with cluster labels.
+
+```{r}
+scater::plotUMAP(sce, colour_by = "celltype_cluster")
+```
+
+### MetaCell-ish
```

Comment:
This feels potentially more of an exercise to me? But, it's pretty quick either way.

Reply:
I think I would rather keep this than the full cluster classification... This may be a thing that we skip if time becomes an issue.